File: pages/login_page.py
"""Page Object for the Sauce Demo Login page."""

from pages.base_page import BasePage
from utils.logger import get_logger

logger = get_logger("login_page")


class LoginPage(BasePage):
    """Encapsulates locators and actions for the login page."""

    # ── XPath Locators ───────────────────────────────────────
    USERNAME_INPUT = "//input[@id='user-name']"
    PASSWORD_INPUT = "//input[@id='password']"
    LOGIN_BUTTON = "//input[@id='login-button']"
    ERROR_MESSAGE = "//h3[@data-test='error']"
    LOGIN_LOGO = "//div[@class='login_logo']"

    def open(self):
        """Navigate to the login page."""
        logger.info("Opening login page")
        self.navigate("/")
        return self

    def enter_username(self, username: str):
        self.fill(self.USERNAME_INPUT, username, step_name="enter_username")
        return self

    def enter_password(self, password: str):
        self.fill(self.PASSWORD_INPUT, password, step_name="enter_password")
        return self

    def click_login(self):
        self.click(self.LOGIN_BUTTON, step_name="click_login")
        return self

    def login(self, username: str, password: str):
        """Full login flow."""
        logger.info("Logging in as: %s", username)
        self.enter_username(username)
        self.enter_password(password)
        self.click_login()
        return self

    def get_error_message(self) -> str:
        return self.get_text(self.ERROR_MESSAGE)

    def is_error_displayed(self) -> bool:
        return self.is_visible(self.ERROR_MESSAGE)

    def is_login_page_loaded(self) -> bool:
        return self.is_visible(self.LOGIN_LOGO)

File: src/main/java/com/framework/pages/LoginPage.java
package com.framework.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Page Object for the SauceDemo Login page.
 * All locators use XPath as preferred strategy.
 */
public class LoginPage extends BasePage {

    // ===== Locators (XPath) =====
    private final By usernameField = By.xpath("//input[@id='user-name']");
    private final By passwordField = By.xpath("//input[@id='password']");
    private final By loginButton = By.xpath("//input[@id='login-button']");
    private final By errorMessage = By.xpath("//div[@class='error-message-container error']//h3");
    private final By loginLogo = By.xpath("//div[@class='login_logo']");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    // ===== Page Actions =====

    public void navigateToLoginPage(String url) {
        logger.info("Navigating to login page: {}", url);
        navigateTo(url);
    }

    public void enterUsername(String username) {
        logger.info("Entering username: {}", username);
        type(usernameField, username);
    }

    public void enterPassword(String password) {
        logger.info("Entering password");
        type(passwordField, password);
    }

    public void clickLoginButton() {
        logger.info("Clicking login button");
        click(loginButton);
    }

    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLoginButton();
    }

    public String getErrorMessage() {
        String error = getText(errorMessage);
        logger.info("Error message: {}", error);
        return error;
    }

    public boolean isErrorMessageDisplayed() {
        return isDisplayed(errorMessage);
    }

    public boolean isLoginPageDisplayed() {
        return isDisplayed(loginLogo);
    }

    public boolean isLoginButtonVisible() {
        return isDisplayed(loginButton);
    }
}

